File: examples/canonical-analytics/python/src/analytics_workbench/analytics/events.py
from dataclasses import dataclass
from typing import List
from datetime import datetime
from ..data.models import Event
from ..filters import FilterState

@dataclass
class EventCountSeries:
    dates: list[str]
    counts: list[int]
    by_environment: dict[str, list[int]]

def aggregate_event_counts(events: list[Event], filters: FilterState) -> EventCountSeries:
    import pandas as pd
    
    # filter
    filtered = []
    for e in events:
        d = e.timestamp[:10]
        if filters.date_from and d < filters.date_from.isoformat(): continue
        if filters.date_to and d > filters.date_to.isoformat(): continue
        if filters.environment and e.environment not in filters.environment: continue
        filtered.append(e)
        
    if not filtered:
        return EventCountSeries(dates=[], counts=[], by_environment={})
        
    df = pd.DataFrame([{"date": e.timestamp[:10], "environment": e.environment, "count": e.count} for e in filtered])
    
    dates = sorted(df["date"].unique().tolist())
    by_env = {}
    
    for env in df["environment"].unique():
        env_df = df[df["environment"] == env]
        counts = env_df.groupby("date")["count"].sum().reindex(dates, fill_value=0).tolist()
        by_env[env] = counts
        
    total_counts = df.groupby("date")["count"].sum().reindex(dates, fill_value=0).tolist()
    
    return EventCountSeries(
        dates=dates,
        counts=total_counts,
        by_environment=by_env
    )

File: examples/canonical-analytics/elixir/analytics_workbench/lib/analytics_workbench/analytics/events.ex
defmodule AnalyticsWorkbench.Analytics.Events do
  alias AnalyticsWorkbench.Filters

  def aggregate_event_counts(events, %Filters{} = filters) do
    filtered = apply_filters(events, filters)

    # Group by date and environment
    grouped = Enum.group_by(filtered, fn e ->
      String.slice(e.timestamp, 0, 10)
    end)

    dates = Map.keys(grouped) |> Enum.sort()

    by_environment = filtered
      |> Enum.group_by(& &1.environment)
      |> Map.new(fn {env, env_events} ->
        env_counts = dates |> Enum.map(fn d ->
          env_events
          |> Enum.filter(& String.starts_with?(&1.timestamp, d))
          |> Enum.map(& &1.count)
          |> Enum.sum()
        end)
        {to_string(env), env_counts}
      end)

    counts = dates |> Enum.map(fn d ->
      filtered
      |> Enum.filter(& String.starts_with?(&1.timestamp, d))
      |> Enum.map(& &1.count)
      |> Enum.sum()
    end)

    %{dates: dates, counts: counts, by_environment: by_environment}
  end

  defp apply_filters(events, filters) do
    events
    |> filter_by_date(filters)
    |> filter_by_multi(:service, filters.services)
    |> filter_by_multi(:environment, filters.environment)
  end

  defp filter_by_date(events, %{date_from: from, date_to: to}) do
    events
    |> Enum.filter(fn e ->
      (is_nil(from) or from == "" or String.slice(e.timestamp, 0, 10) >= from) and
      (is_nil(to) or to == "" or String.slice(e.timestamp, 0, 10) <= to)
    end)
  end

  defp filter_by_multi(events, _key, []), do: events
  defp filter_by_multi(events, key, values) do
    Enum.filter(events, &(to_string(Map.get(&1, key)) in values))
  end
end
